### PR TITLE
Seesaws

### PR DESCRIPTION
**Pull request by jaudick**
https://github.com/NextCenturyCorporation/MCS/pull/391
New solution: Shoot a raycast with gravity force downward from the agent.

The seesaw config property locks the XZ position and YZ rotation of the seesaw rigidbody. It ignores collisions with the agents ground collider. For seesaw pivot points, enable seesaw and set that object to kinematic so that it ignores agent ground collider collisions but will not move.
**Seesaw ---**
Position
X: Locked, Y: Unlocked, Z: Locked
Rotation
X: Unlocked, Y: Locked, Z: Locked

### CODE REVIEW

**Review by kdrumm — APPROVED** [review]

**Comment by ThomasSchellenbergNextCentury** [comment]
@jaudick 

> This means collisions with physics simulated game objects are not consistent with scenes where the agent is kinematic because the agent is not an unstoppable force anymore.

Would you mind elaborating on this point, and providing a few examples? My inclination is that the collision behavior should be consistent across ALL scenes (i.e. the agent's body should no longer be kinematic in ALL interactive scenes), but I want to better understand the ramifications of doing so first. We don't want the AIs to detect some subtle difference in the way the agent interacts with the environment (via collisions) that might cause confusion.

File: unity/Assets/Scripts/MCSMain.cs [hunk 1]
```diff
@@ -69,6 +69,7 @@ public class MCSMain : MonoBehaviour {
     private static float WALL_RIGHT_POSITION_X = 5.25f;
     private static float WALL_SCALE_Y = 3.0f;
     private static Vector3 DEFAULT_ROOM_DIMENSIONS = new Vector3(10, 3, 10);
+    private static float SEESAW_MAX_DEPTH = 0.035f; //any seesaw thicker than this causes collision problems
```
Comment (ThomasSchellenbergNextCentury) [inline unity/Assets/Scripts/MCSMain.cs:72]: Out of curiosity, what kind of collision problems?

File: unity/Assets/Scripts/MCSMain.cs [hunk 2]
```diff
@@ -1796,6 +1813,11 @@ public class MCSConfigScene {
     public bool observation; // deprecated; please use intuitivePhysics
     public bool screenshot;
     public bool isometric;
+    //A seesaw scene makes the agent's rigidbody not kinematic. 
+    //This means collisions with physics simulated game objects are *not* consistent with scenes where the agent *is* kinematic because the agent is not an unstoppable force anymore. 
+    //Agent rigidbody contraints are position locked on XZ. XYZ rotations are also locked. The Y position rigidbody constraint is unlocked which simulates gravity when on a seesaw.
+    //The agent can still walk into objects and push them, but there is a push-back force dependent on the agents velocity and mass.
+    public bool seesawScene;
```
Comment (ThomasSchellenbergNextCentury) [inline unity/Assets/Scripts/MCSMain.cs:1820]: Is there a chance we could rename this property to something more generic; maybe something like `kinematicAgent`?

File: unity/Assets/Scripts/PhysicsRemoteFPSAgentController.cs [hunk 3]
```diff
@@ -4249,7 +4249,8 @@ protected void SetStateOfObjectsThatDontHaveAnimationTime(SetObjectStates SetObj
             //if receptacle can open, check that it's open before placing. Can't place objects in something that is closed!
 
             if (targetReceptacle.DoesThisObjectHaveThisSecondaryProperty(SimObjSecondaryProperty.CanOpen)) {
-                if (!targetReceptacle.GetComponent<CanOpen_Object>().isOpen) {
+                CanOpen_Object canOpenObject = targetReceptacle.GetComponent<CanOpen_Object>();
+                if (canOpenObject != null && !canOpenObject.isOpen) {
```
Comment (ThomasSchellenbergNextCentury) [inline unity/Assets/Scripts/PhysicsRemoteFPSAgentController.cs:4253]: Out of curiosity, why was this change needed?

File: unity/Assets/Scripts/SimObjPhysics.cs [hunk 4]
```diff
@@ -109,6 +109,9 @@ public class SimObjPhysics : MonoBehaviour, SimpleSimObj
     public bool IsDirtyable;
     public bool IsCookable;
     public bool IsSliceable;
+    //IsSeesaw locks the XZ position and YZ rotation of the rigidbody. It ignores collisions with the agents ground collider.
+    //For seesaw pivot points, enable IsSeesaw and set that object to kinematic so that it ignores agent ground collider collisions but will not move.
+    public bool IsSeesaw;
```
Comment (ThomasSchellenbergNextCentury) [inline unity/Assets/Scripts/SimObjPhysics.cs:114]: Would you mind adding a note about this to the Changelog in our README?

File: unity/Assets/Scripts/PhysicsRemoteFPSAgentController.cs (hunk 3 above)
Comment (jaudick) [reply unity/Assets/Scripts/PhysicsRemoteFPSAgentController.cs:4253]: Putting an object on top of the changing table which is doesnt open or close was causing issues because it was calling .isOpen on null. This fixed it.

File: unity/Assets/Scripts/MCSMain.cs (hunk 1 above)
Comment (jaudick) [reply unity/Assets/Scripts/MCSMain.cs:72]: With the way `MatchAgentHeightToStructureBelow` works the agent can't consistently mount a seesaw with a greater depth and its collider would clip into it causing the seesaw to behave unexpectedly

File: unity/Assets/Scripts/MCSMain.cs (hunk 2 above)
Comment (jaudick) [reply unity/Assets/Scripts/MCSMain.cs:1820]: Yes i can do that

File: unity/Assets/Scripts/SimObjPhysics.cs (hunk 4 above)
Comment (jaudick) [reply unity/Assets/Scripts/SimObjPhysics.cs:114]: I will get on that

**Comment by jaudick** [comment]
@ThomasSchellenbergNextCentury this branch with the new solution is ready to be reviewed again. There are updated integration tests on the mcs branch.

File: unity/Assets/Scripts/MCSController.cs [hunk 5]
```diff
@@ -883,7 +885,8 @@ public override void RotateLook(ServerAction response)
 
         //raycast to traverse structures at anything <= 45 degree angle incline
         if (Physics.SphereCast(origin, AGENT_RADIUS, Vector3.down, out hit, Mathf.Infinity, layerMask, QueryTriggerInteraction.Ignore) &&
-            (hit.transform.GetComponent<StructureObject>() != null)) {
+            ((hit.transform.GetComponent<StructureObject>() != null) || (hit.transform.GetComponent<SimObjPhysics>() != null && hit.transform.GetComponent<SimObjPhysics>().IsSeesaw))) {
+            hit.rigidbody.AddForceAtPosition(Vector3.down * GRAVITY_FORCE * GetComponent<Rigidbody>().mass, hit.point);
```
Comment (ThomasSchellenbergNextCentury) [inline unity/Assets/Scripts/MCSController.cs:889]: Why are we adding a downward force here? Is this because we're now keeping the agent always kinematic?

File: unity/Assets/Scripts/MCSController.cs [hunk 6]
```diff
@@ -46,6 +46,8 @@ public class MCSController : PhysicsRemoteFPSAgentController {
 
     //this is not the capsule radius, this is the radius of the x and z bounds of the agent.
     public static float AGENT_RADIUS = 0.12f;
+    public static float GRAVITY_FORCE = 9.81f;
```
Comment (ThomasSchellenbergNextCentury) [inline unity/Assets/Scripts/MCSController.cs:49]: Shouldn't we be able to derive this number from Unity? Like `Physics.gravity` or something?

File: unity/Assets/Scripts/MCSController.cs (hunk 5 above)
Comment (jaudick) [reply unity/Assets/Scripts/MCSController.cs:889]: Yes now the agent is always kinematic

File: unity/Assets/Scripts/MCSController.cs (hunk 6 above)
Comment (jaudick) [reply unity/Assets/Scripts/MCSController.cs:49]: Yes that would work too

Comment (jaudick) [reply unity/Assets/Scripts/MCSController.cs]: @ThomasSchellenbergNextCentury updated with latest commit
